File: frameworks/graphql-go/internal/schema/schema.go
package schema

import (
	"context"

	"github.com/benchmark/graphql-go/internal/dataloader"
	"github.com/benchmark/graphql-go/internal/db"
	"github.com/benchmark/graphql-go/internal/model"
	"github.com/graphql-go/graphql"
)

// Schema is the GraphQL schema
var Schema graphql.Schema

func init() {
	// Forward declarations for recursive types
	var userType *graphql.Object
	var postType *graphql.Object
	var commentType *graphql.Object

	// Comment type
	commentType = graphql.NewObject(graphql.ObjectConfig{
		Name: "Comment",
		Fields: graphql.FieldsThunk(func() graphql.Fields {
			return graphql.Fields{
				"id": &graphql.Field{
					Type: graphql.NewNonNull(graphql.ID),
				},
				"content": &graphql.Field{
					Type: graphql.NewNonNull(graphql.String),
				},
				"author": &graphql.Field{
					Type: userType,
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						comment := p.Source.(*model.Comment)
						if comment.AuthorID == nil {
							return nil, nil
						}
						loaders := dataloader.GetLoaders(p.Context)
						return loaders.UserLoader.Load(p.Context, *comment.AuthorID)()
					},
				},
				"post": &graphql.Field{
					Type: postType,
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						comment := p.Source.(*model.Comment)
						if comment.PostID == nil {
							return nil, nil
						}
						loaders := dataloader.GetLoaders(p.Context)
						return loaders.PostLoader.Load(p.Context, *comment.PostID)()
					},
				},
			}
		}),
	})

	// Post type
	postType = graphql.NewObject(graphql.ObjectConfig{
		Name: "Post",
		Fields: graphql.FieldsThunk(func() graphql.Fields {
			return graphql.Fields{
				"id": &graphql.Field{
					Type: graphql.NewNonNull(graphql.ID),
				},
				"title": &graphql.Field{
					Type: graphql.NewNonNull(graphql.String),
				},
				"content": &graphql.Field{
					Type: graphql.String,
				},
				"createdAt": &graphql.Field{
					Type: graphql.NewNonNull(graphql.String),
				},
				"author": &graphql.Field{
					Type: graphql.NewNonNull(userType),
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						post := p.Source.(*model.Post)
						loaders := dataloader.GetLoaders(p.Context)
						return loaders.UserLoader.Load(p.Context, post.AuthorID)()
					},
				},
				"comments": &graphql.Field{
					Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(commentType))),
					Args: graphql.FieldConfigArgument{
						"limit": &graphql.ArgumentConfig{
							Type:         graphql.Int,
							DefaultValue: 50,
						},
					},
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						post := p.Source.(*model.Post)
						limit := 50
						if l, ok := p.Args["limit"].(int); ok {
							limit = l
						}

						loaders := dataloader.GetLoaders(p.Context)
						comments, err := loaders.CommentsByPostLoader.Load(p.Context, post.ID)()
						if err != nil {
							return nil, err
						}

						if len(comments) > limit {
							comments = comments[:limit]
						}
						return comments, nil
					},
				},
			}
		}),
	})

	// User type
	userType = graphql.NewObject(graphql.ObjectConfig{
		Name: "User",
		Fields: graphql.FieldsThunk(func() graphql.Fields {
			return graphql.Fields{
				"id": &graphql.Field{
					Type: graphql.NewNonNull(graphql.ID),
				},
				"username": &graphql.Field{
					Type: graphql.NewNonNull(graphql.String),
				},
				"fullName": &graphql.Field{
					Type: graphql.String,
				},
				"bio": &graphql.Field{
					Type: graphql.String,
				},
				"posts": &graphql.Field{
					Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(postType))),
					Args: graphql.FieldConfigArgument{
						"limit": &graphql.ArgumentConfig{
							Type:         graphql.Int,
							DefaultValue: 50,
						},
					},
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						user := p.Source.(*model.User)
						limit := 50
						if l, ok := p.Args["limit"].(int); ok {
							limit = l
						}

						loaders := dataloader.GetLoaders(p.Context)
						posts, err := loaders.PostsByAuthorLoader.Load(p.Context, user.ID)()
						if err != nil {
							return nil, err
						}

						if len(posts) > limit {
							posts = posts[:limit]
						}
						return posts, nil
					},
				},
				"followers": &graphql.Field{
					Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(userType))),
					Args: graphql.FieldConfigArgument{
						"limit": &graphql.ArgumentConfig{
							Type:         graphql.Int,
							DefaultValue: 50,
						},
					},
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						// Followers relationship not implemented in benchmark schema
						return []*model.User{}, nil
					},
				},
				"following": &graphql.Field{
					Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(userType))),
					Args: graphql.FieldConfigArgument{
						"limit": &graphql.ArgumentConfig{
							Type:         graphql.Int,
							DefaultValue: 50,
						},
					},
					Resolve: func(p graphql.ResolveParams) (interface{}, error) {
						// Following relationship not implemented in benchmark schema
						return []*model.User{}, nil
					},
				},
			}
		}),
	})

	// Query type
	queryType := graphql.NewObject(graphql.ObjectConfig{
		Name: "Query",
		Fields: graphql.Fields{
			"ping": &graphql.Field{
				Type: graphql.NewNonNull(graphql.String),
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					return "pong", nil
				},
			},
			"user": &graphql.Field{
				Type: userType,
				Args: graphql.FieldConfigArgument{
					"id": &graphql.ArgumentConfig{
						Type: graphql.NewNonNull(graphql.ID),
					},
				},
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					id := p.Args["id"].(string)
					loaders := dataloader.GetLoaders(p.Context)
					return loaders.UserLoader.Load(p.Context, id)()
				},
			},
			"users": &graphql.Field{
				Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(userType))),
				Args: graphql.FieldConfigArgument{
					"limit": &graphql.ArgumentConfig{
						Type:         graphql.Int,
						DefaultValue: 10,
					},
				},
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					limit := 10
					if l, ok := p.Args["limit"].(int); ok {
						limit = l
					}

					rows, err := db.Pool.Query(p.Context, `
						SELECT id, username, full_name, bio
						FROM benchmark.tb_user
						ORDER BY created_at DESC
						LIMIT $1
					`, limit)
					if err != nil {
						return nil, err
					}
					defer rows.Close()

					var users []*model.User
					for rows.Next() {
						var u model.User
						var fullName, bio *string
						if err := rows.Scan(&u.ID, &u.Username, &fullName, &bio); err != nil {
							continue
						}
						u.FullName = fullName
						u.Bio = bio
						users = append(users, &u)
					}

					if users == nil {
						users = []*model.User{}
					}
					return users, nil
				},
			},
			"post": &graphql.Field{
				Type: postType,
				Args: graphql.FieldConfigArgument{
					"id": &graphql.ArgumentConfig{
						Type: graphql.NewNonNull(graphql.ID),
					},
				},
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					id := p.Args["id"].(string)
					loaders := dataloader.GetLoaders(p.Context)
					return loaders.PostLoader.Load(p.Context, id)()
				},
			},
			"posts": &graphql.Field{
				Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(postType))),
				Args: graphql.FieldConfigArgument{
					"limit": &graphql.ArgumentConfig{
						Type:         graphql.Int,
						DefaultValue: 10,
					},
				},
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					limit := 10
					if l, ok := p.Args["limit"].(int); ok {
						limit = l
					}

					rows, err := db.Pool.Query(p.Context, `
						SELECT p.id, u.id as author_id, p.title, p.content, p.created_at
						FROM benchmark.tb_post p
						JOIN benchmark.tb_user u ON p.fk_author = u.pk_user
						ORDER BY p.created_at DESC
						LIMIT $1
					`, limit)
					if err != nil {
						return nil, err
					}
					defer rows.Close()

					var posts []*model.Post
					for rows.Next() {
						var p model.Post
						var content *string
						var createdAt string
						if err := rows.Scan(&p.ID, &p.AuthorID, &p.Title, &content, &createdAt); err != nil {
							continue
						}
						p.Content = content
						p.CreatedAt = createdAt
						posts = append(posts, &p)
					}

					if posts == nil {
						posts = []*model.Post{}
					}
					return posts, nil
				},
			},
		},
	})

	// Mutation type
	mutationType := graphql.NewObject(graphql.ObjectConfig{
		Name: "Mutation",
		Fields: graphql.Fields{
			"updateUser": &graphql.Field{
				Type: userType,
				Args: graphql.FieldConfigArgument{
					"id": &graphql.ArgumentConfig{
						Type: graphql.NewNonNull(graphql.ID),
					},
					"fullName": &graphql.ArgumentConfig{
						Type: graphql.String,
					},
					"bio": &graphql.ArgumentConfig{
						Type: graphql.String,
					},
				},
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					id := p.Args["id"].(string)
					fullName, hasFullName := p.Args["fullName"].(string)
					bio, hasBio := p.Args["bio"].(string)

					query := "UPDATE benchmark.tb_user SET updated_at = NOW()"
					args := []interface{}{}
					argIdx := 1

					if hasFullName {
						query += ", full_name = $" + string(rune('0'+argIdx))
						args = append(args, fullName)
						argIdx++
					}
					if hasBio {
						query += ", bio = $" + string(rune('0'+argIdx))
						args = append(args, bio)
						argIdx++
					}

					query += " WHERE id = $" + string(rune('0'+argIdx))
					args = append(args, id)

					_, err := db.Pool.Exec(p.Context, query, args...)
					if err != nil {
						return nil, err
					}

					// Return updated user via dataloader
					loaders := dataloader.GetLoaders(p.Context)
					loaders.UserLoader.Clear(p.Context, id)
					return loaders.UserLoader.Load(p.Context, id)()
				},
			},
			"updatePost": &graphql.Field{
				Type: postType,
				Args: graphql.FieldConfigArgument{
					"id": &graphql.ArgumentConfig{
						Type: graphql.NewNonNull(graphql.ID),
					},
					"title": &graphql.ArgumentConfig{
						Type: graphql.String,
					},
					"content": &graphql.ArgumentConfig{
						Type: graphql.String,
					},
				},
				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
					id := p.Args["id"].(string)
					title, hasTitle := p.Args["title"].(string)
					content, hasContent := p.Args["content"].(string)

					query := "UPDATE benchmark.tb_post SET updated_at = NOW()"
					args := []interface{}{}
					argIdx := 1

					if hasTitle {
						query += ", title = $" + string(rune('0'+argIdx))
						args = append(args, title)
						argIdx++
					}
					if hasContent {
						query += ", content = $" + string(rune('0'+argIdx))
						args = append(args, content)
						argIdx++
					}

					query += " WHERE id = $" + string(rune('0'+argIdx))
					args = append(args, id)

					_, err := db.Pool.Exec(p.Context, query, args...)
					if err != nil {
						return nil, err
					}

					// Return updated post via dataloader
					loaders := dataloader.GetLoaders(p.Context)
					loaders.PostLoader.Clear(p.Context, id)
					return loaders.PostLoader.Load(p.Context, id)()
				},
			},
		},
	})

	// Build schema
	var err error
	Schema, err = graphql.NewSchema(graphql.SchemaConfig{
		Query:    queryType,
		Mutation: mutationType,
	})
	if err != nil {
		panic(err)
	}
}

// ExecuteQuery executes a GraphQL query
func ExecuteQuery(ctx context.Context, query string, variables map[string]interface{}) *graphql.Result {
	return graphql.Do(graphql.Params{
		Schema:         Schema,
		RequestString:  query,
		VariableValues: variables,
		Context:        ctx,
	})
}

File: frameworks/juniper/src/schema.rs
use juniper::{graphql_object, EmptySubscription, FieldResult, GraphQLInputObject, RootNode, ID};
use uuid::Uuid;

use crate::db::Database;
use crate::loaders::Loaders;
use crate::models::{Comment, Post, User};

/// GraphQL context containing database and loaders
pub struct Context {
    pub db: Database,
    pub loaders: Loaders,
}

impl juniper::Context for Context {}

// GraphQL object implementations

#[graphql_object(context = Context)]
impl User {
    fn id(&self) -> ID {
        ID::new(self.id.to_string())
    }

    fn username(&self) -> &str {
        &self.username
    }

    fn full_name(&self) -> Option<&str> {
        self.full_name.as_deref()
    }

    fn bio(&self) -> Option<&str> {
        self.bio.as_deref()
    }

    fn created_at(&self) -> String {
        self.created_at.to_rfc3339()
    }

    async fn posts(&self, context: &Context, limit: Option<i32>) -> FieldResult<Vec<Post>> {
        let limit = limit.unwrap_or(50).min(50);
        let posts = context
            .loaders
            .posts_by_author_loader
            .load(self.pk_user, limit)
            .await
            .map_err(|e| juniper::FieldError::new(e, juniper::Value::null()))?;
        Ok(posts)
    }

    async fn followers(&self, _context: &Context, _limit: Option<i32>) -> FieldResult<Vec<User>> {
        // Followers relationship not implemented in benchmark schema
        Ok(vec![])
    }

    async fn following(&self, _context: &Context, _limit: Option<i32>) -> FieldResult<Vec<User>> {
        // Following relationship not implemented in benchmark schema
        Ok(vec![])
    }
}

#[graphql_object(context = Context)]
impl Post {
    fn id(&self) -> ID {
        ID::new(self.id.to_string())
    }

    fn title(&self) -> &str {
        &self.title
    }

    fn content(&self) -> Option<&str> {
        self.content.as_deref()
    }

    fn created_at(&self) -> String {
        self.created_at.to_rfc3339()
    }

    async fn author(&self, context: &Context) -> FieldResult<User> {
        context
            .loaders
            .user_loader
            .load(self.fk_author)
            .await
            .map_err(|e| juniper::FieldError::new(e, juniper::Value::null()))?
            .ok_or_else(|| juniper::FieldError::new("Author not found", juniper::Value::null()))
    }

    async fn comments(&self, context: &Context, limit: Option<i32>) -> FieldResult<Vec<Comment>> {
        let limit = limit.unwrap_or(50).min(50);
        let comments = context
            .loaders
            .comments_by_post_loader
            .load(self.pk_post, limit)
            .await
            .map_err(|e| juniper::FieldError::new(e, juniper::Value::null()))?;
        Ok(comments)
    }
}

#[graphql_object(context = Context)]
impl Comment {
    fn id(&self) -> ID {
        ID::new(self.id.to_string())
    }

    fn content(&self) -> &str {
        &self.content
    }

    fn created_at(&self) -> String {
        self.created_at.to_rfc3339()
    }

    async fn author(&self, context: &Context) -> FieldResult<User> {
        context
            .loaders
            .user_loader
            .load(self.fk_author)
            .await
            .map_err(|e| juniper::FieldError::new(e, juniper::Value::null()))?
            .ok_or_else(|| juniper::FieldError::new("Author not found", juniper::Value::null()))
    }

    async fn post(&self, context: &Context) -> FieldResult<Post> {
        context
            .loaders
            .post_loader
            .load_by_pk(self.fk_post)
            .await
            .map_err(|e| juniper::FieldError::new(e, juniper::Value::null()))?
            .ok_or_else(|| juniper::FieldError::new("Post not found", juniper::Value::null()))
    }
}

// Query root

pub struct QueryRoot;

#[graphql_object(context = Context)]
impl QueryRoot {
    fn ping() -> &'static str {
        "pong"
    }

    async fn user(context: &Context, id: ID) -> FieldResult<Option<User>> {
        let user_id = Uuid::parse_str(&id.to_string())
            .map_err(|e| juniper::FieldError::new(format!("Invalid UUID: {}", e), juniper::Value::null()))?;

        let client = context.db.pool().get().await
            .map_err(|e| juniper::FieldError::new(format!("DB error: {}", e), juniper::Value::null()))?;

        let row = client
            .query_opt(
                "SELECT id, pk_user, username, full_name, bio, created_at
                 FROM benchmark.tb_user WHERE id = $1",
                &[&user_id],
            )
            .await
            .map_err(|e| juniper::FieldError::new(format!("Query error: {}", e), juniper::Value::null()))?;

        Ok(row.map(|r| User {
            id: r.get(0),
            pk_user: r.get(1),
            username: r.get(2),
            full_name: r.get(3),
            bio: r.get(4),
            created_at: r.get(5),
        }))
    }

    async fn users(context: &Context, limit: Option<i32>) -> FieldResult<Vec<User>> {
        let limit = limit.unwrap_or(10).min(100) as i64;

        let client = context.db.pool().get().await
            .map_err(|e| juniper::FieldError::new(format!("DB error: {}", e), juniper::Value::null()))?;

        let rows = client
            .query(
                "SELECT id, pk_user, username, full_name, bio, created_at
                 FROM benchmark.tb_user
                 ORDER BY created_at DESC
                 LIMIT $1",
                &[&limit],
            )
            .await
            .map_err(|e| juniper::FieldError::new(format!("Query error: {}", e), juniper::Value::null()))?;

        Ok(rows
            .iter()
            .map(|r| User {
                id: r.get(0),
                pk_user: r.get(1),
                username: r.get(2),
                full_name: r.get(3),
                bio: r.get(4),
                created_at: r.get(5),
            })
            .collect())
    }

    async fn post(context: &Context, id: ID) -> FieldResult<Option<Post>> {
        let post_id = Uuid::parse_str(&id.to_string())
            .map_err(|e| juniper::FieldError::new(format!("Invalid UUID: {}", e), juniper::Value::null()))?;

        context
            .loaders
            .post_loader
            .load_by_uuid(post_id)
            .await
            .map_err(|e| juniper::FieldError::new(e, juniper::Value::null()))
    }

    async fn posts(context: &Context, limit: Option<i32>) -> FieldResult<Vec<Post>> {
        let limit = limit.unwrap_or(10).min(100) as i64;

        let client = context.db.pool().get().await
            .map_err(|e| juniper::FieldError::new(format!("DB error: {}", e), juniper::Value::null()))?;

        let rows = client
            .query(
                "SELECT id, pk_post, title, content, fk_author, created_at
                 FROM benchmark.tb_post
                 ORDER BY created_at DESC
                 LIMIT $1",
                &[&limit],
            )
            .await
            .map_err(|e| juniper::FieldError::new(format!("Query error: {}", e), juniper::Value::null()))?;

        Ok(rows
            .iter()
            .map(|r| Post {
                id: r.get(0),
                pk_post: r.get(1),
                title: r.get(2),
                content: r.get(3),
                fk_author: r.get(4),
                created_at: r.get(5),
            })
            .collect())
    }

    async fn comments(context: &Context, limit: Option<i32>) -> FieldResult<Vec<Comment>> {
        let limit = limit.unwrap_or(20).min(100) as i64;

        let client = context.db.pool().get().await
            .map_err(|e| juniper::FieldError::new(format!("DB error: {}", e), juniper::Value::null()))?;

        let rows = client
            .query(
                "SELECT id, pk_comment, content, fk_post, fk_author, created_at
                 FROM benchmark.tb_comment
                 ORDER BY created_at DESC
                 LIMIT $1",
                &[&limit],
            )
            .await
            .map_err(|e| juniper::FieldError::new(format!("Query error: {}", e), juniper::Value::null()))?;

        Ok(rows
            .iter()
            .map(|r| Comment {
                id: r.get(0),
                pk_comment: r.get(1),
                content: r.get(2),
                fk_post: r.get(3),
                fk_author: r.get(4),
                created_at: r.get(5),
            })
            .collect())
    }
}

// Mutation root

#[derive(GraphQLInputObject)]
pub struct UpdateUserInput {
    pub full_name: Option<String>,
    pub bio: Option<String>,
}

#[derive(GraphQLInputObject)]
pub struct UpdatePostInput {
    pub title: Option<String>,
    pub content: Option<String>,
}

pub struct MutationRoot;

#[graphql_object(context = Context)]
impl MutationRoot {
    async fn update_user(context: &Context, id: ID, input: UpdateUserInput) -> FieldResult<User> {
        let user_id = Uuid::parse_str(&id.to_string())
            .map_err(|e| juniper::FieldError::new(format!("Invalid UUID: {}", e), juniper::Value::null()))?;

        let client = context.db.pool().get().await
            .map_err(|e| juniper::FieldError::new(format!("DB error: {}", e), juniper::Value::null()))?;

        // Build update query dynamically
        let mut query = String::from("UPDATE benchmark.tb_user SET updated_at = NOW()");
        let mut params: Vec<&(dyn tokio_postgres::types::ToSql + Sync)> = vec![];
        let mut param_idx = 1;

        if let Some(ref full_name) = input.full_name {
            query.push_str(&format!(", full_name = ${}", param_idx));
            params.push(full_name);
            param_idx += 1;
        }

        if let Some(ref bio) = input.bio {
            query.push_str(&format!(", bio = ${}", param_idx));
            params.push(bio);
            param_idx += 1;
        }

        query.push_str(&format!(" WHERE id = ${}", param_idx));
        params.push(&user_id);

        if input.full_name.is_some() || input.bio.is_some() {
            client
                .execute(&query, &params[..])
                .await
                .map_err(|e| juniper::FieldError::new(format!("Update error: {}", e), juniper::Value::null()))?;
        }

        // Return updated user
        let row = client
            .query_one(
                "SELECT id, pk_user, username, full_name, bio, created_at
                 FROM benchmark.tb_user WHERE id = $1",
                &[&user_id],
            )
            .await
            .map_err(|e| juniper::FieldError::new(format!("Query error: {}", e), juniper::Value::null()))?;

        Ok(User {
            id: row.get(0),
            pk_user: row.get(1),
            username: row.get(2),
            full_name: row.get(3),
            bio: row.get(4),
            created_at: row.get(5),
        })
    }

    async fn update_post(context: &Context, id: ID, input: UpdatePostInput) -> FieldResult<Post> {
        let post_id = Uuid::parse_str(&id.to_string())
            .map_err(|e| juniper::FieldError::new(format!("Invalid UUID: {}", e), juniper::Value::null()))?;

        let client = context.db.pool().get().await
            .map_err(|e| juniper::FieldError::new(format!("DB error: {}", e), juniper::Value::null()))?;

        // Build update query dynamically
        let mut query = String::from("UPDATE benchmark.tb_post SET updated_at = NOW()");
        let mut params: Vec<&(dyn tokio_postgres::types::ToSql + Sync)> = vec![];
        let mut param_idx = 1;

        if let Some(ref title) = input.title {
            query.push_str(&format!(", title = ${}", param_idx));
            params.push(title);
            param_idx += 1;
        }

        if let Some(ref content) = input.content {
            query.push_str(&format!(", content = ${}", param_idx));
            params.push(content);
            param_idx += 1;
        }

        query.push_str(&format!(" WHERE id = ${}", param_idx));
        params.push(&post_id);

        if input.title.is_some() || input.content.is_some() {
            client
                .execute(&query, &params[..])
                .await
                .map_err(|e| juniper::FieldError::new(format!("Update error: {}", e), juniper::Value::null()))?;
        }

        // Return updated post
        let row = client
            .query_one(
                "SELECT id, pk_post, title, content, fk_author, created_at
                 FROM benchmark.tb_post WHERE id = $1",
                &[&post_id],
            )
            .await
            .map_err(|e| juniper::FieldError::new(format!("Query error: {}", e), juniper::Value::null()))?;

        Ok(Post {
            id: row.get(0),
            pk_post: row.get(1),
            title: row.get(2),
            content: row.get(3),
            fk_author: row.get(4),
            created_at: row.get(5),
        })
    }
}

// Schema type alias
pub type Schema = RootNode<'static, QueryRoot, MutationRoot, EmptySubscription<Context>>;

pub fn create_schema() -> Schema {
    Schema::new(QueryRoot, MutationRoot, EmptySubscription::new())
}
